### PR TITLE
TIP-609:  Add localisation on channel

### PR DESCRIPTION
**Pull request by momoss**
- [x] add possibility to translate a channel 
- [x] fix impact on imports/exports
- [x] call translatable labels everywhere in the PIM
- [x] add a migration script

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | Y
| Changelog updated                 | IP
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) |
| Migration script                  |
| Tech Doc                          | N

### CODE REVIEW

File: features/datagrid/datagrid_views.feature [hunk 1]
```diff
@@ -130,3 +130,12 @@ Feature: Datagrid views
     And I should see the flash message "Datagrid view successfully removed"
     And I should see the text "Default view"
     But I should not see the text "Sneakers only"
+
+  Scenario: Successfully change grid scope
+    And I filter by "scope" with operator "" and value "Mobile"
+    And I create the view:
```
Comment (skeleton) [inline features/datagrid/datagrid_views.feature]: can you explain why you need to create a view ?

File: features/datagrid/datagrid_views.feature [hunk 2]
```diff
@@ -130,3 +130,12 @@ Feature: Datagrid views
...
+      | new-view-label | Mobile only |
+    Then I should be on the products page
+    And I should see the flash message "Datagrid view successfully created"
+    And I should see the text "Mobile only"
+    And I should see the text "Mobile"
```
Comment (skeleton) [inline features/datagrid/datagrid_views.feature]: so you didn't change the scope finally or I misunderstood something

File: src/Pim/Component/Catalog/Model/ChannelInterface.php [hunk 3]
```diff
@@ -12,7 +13,7 @@
  * @copyright 2014 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
  */
-interface ChannelInterface extends ReferableInterface, VersionableInterface, ChosableInterface
+interface ChannelInterface extends ReferableInterface, VersionableInterface, ChosableInterface, TranslatableInterface
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Model/ChannelInterface.php]: bc break

File: src/Pim/Component/Connector/Validator/Constraints/ChannelValidator.php [hunk 4]
```diff
@@ -19,20 +20,24 @@ class ChannelValidator extends ChoiceValidator
...
+        $this->userContext       = $userContext;
     }
 
     /**
      * {@inheritdoc}
      */
     public function validate($value, Constraint $constraint)
     {
-        $channels = $this->channelRepository->getLabelsIndexedByCode();
+        $channels = $this->channelRepository->getLabelsIndexedByCode($this->userContext->getCurrentLocaleCode());
```
Comment (skeleton) [inline src/Pim/Component/Connector/Validator/Constraints/ChannelValidator.php]: Don't call `$this->userContext->getCurrentLocaleCode()`. It searches in  `Symfony\Component\HttpFoundation\Request` if locale exists, but you don't care about http request as you are in the connector component.
As you don't care about label here, replace `$this->channelRepository->getLabelsIndexedByCode()` by  `$this->channelRepository->getChannelCodes`

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/CompletenessRepository.php [hunk 5]
```diff
@@ -51,7 +51,7 @@ public function __construct(
     /**
      * {@inheritdoc}
      */
-    public function getProductsCountPerChannels()
+    public function getProductsCountPerChannels($userLocale)
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/CompletenessRepository.php]: $localeCode
you don't care here it comes from the user or from anything else

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/CompletenessRepository.php [hunk 6]
```diff
@@ -82,7 +82,7 @@ public function getProductsCountPerChannels()
     /**
      * {@inheritdoc}
      */
-    public function getCompleteProductsCountPerChannels()
+    public function getCompleteProductsCountPerChannels($userLocale)
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/CompletenessRepository.php]: $localeCode
you don't care here it comes from the user or from anything else

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ChannelRepository.php [hunk 7]
```diff
@@ -179,14 +187,19 @@ public function getIdentifierProperties()
     /**
      * {@inheritdoc}
      */
-    public function getLabelsIndexedByCode()
+    public function getLabelsIndexedByCode($userLocaleCode)
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ChannelRepository.php]: $localeCode
you don't care here it comes from the user or from anything else

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CompletenessRepository.php [hunk 8]
```diff
@@ -39,18 +39,21 @@ public function __construct(
     /**
      * {@inheritdoc}
      */
-    public function getProductsCountPerChannels()
+    public function getProductsCountPerChannels($userLocale)
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CompletenessRepository.php]: $localeCode
you don't care here it comes from the user or from anything else

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CompletenessRepository.php [hunk 9]
```diff
@@ -39,18 +39,21 @@ public function __construct(
...
         $sql = <<<SQL
-SELECT ch.label, COUNT(DISTINCT p.id) as total FROM pim_catalog_channel ch
+SELECT t.label, COUNT(DISTINCT p.id) as total FROM pim_catalog_channel ch
+    JOIN pim_catalog_channel_translation t ON t.foreign_key = ch.id
     JOIN %category_table% ca ON ca.root = ch.category_id
     JOIN %category_join_table% cp ON cp.category_id = ca.id
     JOIN %product_table% p ON p.id = cp.product_id
     WHERE p.is_enabled = 1
-    GROUP BY ch.id, ch.label
+    AND t.locale = '%current_user_locale%'
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CompletenessRepository.php]: %locale%

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CompletenessRepository.php [hunk 10]
```diff
@@ -61,10 +64,11 @@ public function getProductsCountPerChannels()
     /**
      * {@inheritdoc}
      */
-    public function getCompleteProductsCountPerChannels()
+    public function getCompleteProductsCountPerChannels($userLocale)
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CompletenessRepository.php]: $localeCode
you don't care here it comes from the user or from anything else

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CompletenessRepository.php [hunk 11]
```diff
@@ -73,9 +77,11 @@ public function getCompleteProductsCountPerChannels()
     LEFT JOIN pim_catalog_completeness co
         ON co.locale_id = lo.id AND co.channel_id = ch.id AND co.product_id = p.id AND co.ratio = 100
     WHERE p.is_enabled = 1
-    GROUP BY ch.id, lo.id, ch.label, lo.code
+    AND t.locale = '%current_user_locale%'
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CompletenessRepository.php]: %locale%

File: src/Pim/Component/Catalog/Localization/ChannelUpdater.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    public function update($channel, array $data, array $options = [])
+    {
+        try {
+            $this->channelUpdater->update($channel, $data, $options);
+
+            if (isset($data['labels']) && $channel instanceof TranslatableInterface) {
+                $this->translatableUpdater->update($channel, $data['labels']);
+            }
+        } catch (\InvalidArgumentException $e) {
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Localization/ChannelUpdater.php]: why do you catch an exception to throw exactly the same?
not exactly the, here you loose some important information (code, stacktrace...)

File: src/Pim/Component/Catalog/Normalizer/Structured/ChannelNormalizer.php [hunk 13]
```diff
@@ -20,6 +20,21 @@ class ChannelNormalizer implements NormalizerInterface
     protected $supportedFormats = ['json', 'xml'];
 
     /**
+     * @var TranslationNormalizer
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ChannelNormalizer.php]: NormalizerInterface

File: src/Pim/Component/Catalog/Repository/ChannelRepositoryInterface.php [hunk 14]
```diff
@@ -68,7 +68,9 @@ public function getChannelCountUsingCurrency(CurrencyInterface $currency);
      * Get channel choices
      * Allow to list channels in an array like array[<code>] = <label>
      *
+     * @param string $userLocaleCode
+     *
      * @return string[]
      */
-    public function getLabelsIndexedByCode();
+    public function getLabelsIndexedByCode($userLocaleCode);
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Repository/ChannelRepositoryInterface.php]: $localeCode

File: src/Pim/Component/Catalog/Repository/CompletenessRepositoryInterface.php [hunk 15]
```diff
@@ -27,9 +27,11 @@
      *        ),
      *    )
      *
+     * @param string $userLocale
+     *
      * @return array
      */
-    public function getProductsCountPerChannels();
+    public function getProductsCountPerChannels($userLocale);
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Repository/CompletenessRepositoryInterface.php]: $localeCode

File: src/Pim/Component/Catalog/Repository/CompletenessRepositoryInterface.php [hunk 16]
```diff
@@ -58,7 +60,9 @@ public function getProductsCountPerChannels();
      *        ),
      *    )
      *
+     * @param string $userLocale
+     *
      * @return array
      */
-    public function getCompleteProductsCountPerChannels();
+    public function getCompleteProductsCountPerChannels($userLocale);
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Repository/CompletenessRepositoryInterface.php]: $localeCode

File: features/channel/create_channel.feature [hunk 17]
```diff
@@ -8,12 +8,12 @@ Feature: Create a channel
...
-      | Category tree | 2014 collection |
-      | Currencies    | EUR             |
-      | Locales       | French          |
+      | Code                    | foo             |
+      | Category tree           | 2014 collection |
+      | Currencies              | EUR             |
+      | Locales                 | French          |
+      | English (United States) | Bar Bar         |
     And I press the "Save" button
     Then I should see the flash message "Channel successfully saved"
```
Comment (skeleton) [inline features/channel/create_channel.feature]: done

File: features/datagrid/datagrid_views.feature (hunk 2 above)
Comment (skeleton) [inline features/datagrid/datagrid_views.feature]: I added `Given I should see the text "Tablet"` to be sure by default we are on table channel

File: src/Pim/Bundle/DashboardBundle/Widget/CompletenessWidget.php [hunk 18]
```diff
@@ -20,14 +21,22 @@ class CompletenessWidget implements WidgetInterface
...
      * @param CompletenessRepositoryInterface $completenessRepo
      * @param LocaleHelper                    $localeHelper
+     * @param UserContext                     $userContext
      */
-    public function __construct(CompletenessRepositoryInterface $completenessRepo, LocaleHelper $localeHelper)
-    {
+    public function __construct(
+        CompletenessRepositoryInterface $completenessRepo,
+        LocaleHelper $localeHelper,
+        UserContext $userContext
```
Comment (skeleton) [inline src/Pim/Bundle/DashboardBundle/Widget/CompletenessWidget.php]: done

File: src/Pim/Component/Catalog/Model/ChannelInterface.php (hunk 3 above)
Comment (skeleton) [inline src/Pim/Component/Catalog/Model/ChannelInterface.php]: done

File: src/Pim/Bundle/CatalogBundle/Resources/config/doctrine/ChannelTranslation.orm.yml [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+            nullable: true
+        locale:
+            type: string
+            length: 20
+    manyToOne:
+        foreignKey:
+            targetEntity: Pim\Component\Catalog\Model\ChannelInterface
+            inversedBy: translations
+            joinColumns:
+                foreign_key:
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Resources/config/doctrine/ChannelTranslation.orm.yml]: why do we call that foreign_key?

File: upgrades/schema/Version_1_7_20161020144045_channel_labels.php [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        $query = sprintf('SELECT id, label FROM %s', self::CHANNEL_TABLE);
+        $stmt = $this->connection->prepare($query);
+        $stmt->execute();
+        $channels = $stmt->fetchAll();
+
+        $this->addSql(<<<SQL
+            CREATE TABLE IF NOT EXISTS `pim_catalog_channel_translation` (
+              `id` INT AUTO_INCREMENT NOT NULL,
+              `foreign_key` int(11) DEFAULT NULL,
```
Comment (jjanvier) [inline upgrades/schema/Version_1_7_20161020144045_channel_labels.php]: why do we call that foreign_key?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/CompletenessRepository.php [hunk 21]
```diff
@@ -82,7 +82,7 @@ public function getProductsCountPerChannels()
     /**
      * {@inheritdoc}
      */
-    public function getCompleteProductsCountPerChannels()
+    public function getCompleteProductsCountPerChannels($localeCode)
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/CompletenessRepository.php]: Here this parameter does not do anything. We need to have the same data returned in Orm and Mysql.

**Comment by SamirBoulil** [comment]
One question on the getCompleteProductsCountPerChannels function. Then GTM :)

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/CompletenessRepository.php (hunk 21 above)
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Repository/CompletenessRepository.php]: in mongo, it's true it's useless, but this param has been added to respect the interface.
in orm, we need it to inject the locale directly in the sql.

File: src/Pim/Bundle/CatalogBundle/Resources/config/doctrine/ChannelTranslation.orm.yml (hunk 19 above)
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Resources/config/doctrine/ChannelTranslation.orm.yml]: I suppose it's a copy paste from ChannelTranslation.orm.yml
I changed it

**Pull request merged**